### ISSUE TITLE
[Feature Request] Direct link to product review.

### ISSUE THREAD

**Issue by 23d1** [owldesign/QARR#13]
What do you think about the possibility of creating a direct link to post a review? Essentially the inputs would be same as the modal for writing a review, with the potential to have some sort of hash or other url format that either pre-fills the name and email or hides those specific fields based on that. Preferrably templatable, so that it's something that can be included among other stuff like product photography or a bespoke page build.

If that was a link that could be sent via email and so forth, that would definitely up the engagement.

**Comment by owldesign** [comment]
Very interesting.. i'll look into this

**Comment by owldesign** [comment]
Ok I've added something for this. Check it out

Read docs for [direct links](https://docs.qarr.tools/campaigns/direct-links/)

This specific input `<input type="hidden" name="directLinkId" value="{{ model.id }}">` will allow for tracking. So if your user submits review/question you'll have record that they have completed submission.

The generated url will check if `currentUser` matches the associated user or admin. If another user gets a hold of the url they will be redirected to homepage if permissions don't check out.


Let me know if this is close to what you were thinking, if you got suggestions or another approach let me know, I can probably extend this further.

**Comment by 23d1** [comment]
Fantastic! Thanks.

Not sure when I'll have time to check it out over the long weekend, but would it be possible to url format for prefilling? Perhaps associate a hash with the email recipient to essentially prefill the email field, and potentially the order? Use cases would be to send out after a couple of weeks to emails associated with orders, that could potentially be guest checkout.

Let me know what you think. :)

**Comment by owldesign** [comment]
Ah I'm not sure im following..
The direct link url that gets generated is hashed,  it has information associated with product and user. 

Example url: `https://yourwebsite.com/qarr/direct/qopdeblkowznowngrqyjvxae`


I could be overthinking this and not understanding you correctly, if there is other ways you can explain it to me I'm sure I can make it happen.

Maybe walk me step by step of the use flow. Eh not sure how to help you help me :)

**Comment by 23d1** [comment]
Ah, gotcha! Sorry, I should have thoroughly read the docs. I guess the questions is; would it work for guest orders, as opposed to only registered and logged in users?

Other than that, in terms of steps;
1. User or guest orders
2. Some time after the fact, there might be a marketing campaign to drive reviews, so an email is sent out to prior orders (maybe not all, perhaps even driving to two different templates for some quick A-B testing)
3. The link that is in the email, with the corresponding hash that would then be associated with the name and email which shows up pre-filled in the template if they exist (`value="{{ order.name }}"` and so on).
4. The review is submitted, and we profit like crazy :)

**Comment by owldesign** [comment]
Starting to make more sense, I think :)

Is it correct to say that you basically want a URL that's simply based off an email address (regardless of user status) and Order element? 

For example `https://yoursite.com/r/?email=vadim@vadim.com&orderId=46` or we can encrypt all that stuff if security is concern `https://yoursite.com/r/?code=somesupersecrethash`

1. Why if order has multiple products, should that url include `&productId=433` too?
2. Will these links be created manually or do we need some tool to auto create a list of links based on some criteria?

We on the right track?

**Comment by 23d1** [comment]
Yes, exactly that. And it could be a link per product. That way one could separate the CTAs per product, or simply just ask for one, then later send for the other and so forth.

Maybe links based on products in order, but definitely a manual way to do it as well.

**Comment by owldesign** [comment]
Ok here's what I got (with examples)

### Sample url pattern

`https://qarrdemo.test/c/r?email=guest1@gmail.com`

### Terminalogy

* `c` - stands for commerce
* `r` - stands for reviews

### Result

This will give you a list of links you can use (this is because your url did not provide all the necessary query params to show you the form). Url that only contains `?email` will fallback to the below template to list all possible links.

![image](https://user-images.githubusercontent.com/438764/60641798-4d814d00-9de2-11e9-91e2-e0bb8e23ded1.png)

### Complete url pattern

If your url pattern contains all the necessary query params, it will display form.

`https://qarrdemo.test/c/r?email=guest1@gmail.com&orderNumber=12a07bb8acc7992e3bd800e9ed42217e&productId=63`

#### Query Params

* `email` - guest/customer email address
* `orderNumber` - order number
* `productId` - product id


#### Template Variables

These variables will be available to your template

* `order` - Commerce order model
* `product` - Commerce product model
* `email` - customer email address
* `name` - customer name (pulled from order details, if not registered user)
* `customer` - User model (if registered customer)

![image](https://user-images.githubusercontent.com/438764/60641936-e021ec00-9de2-11e9-8bec-6ec8a5205227.png)

Ok, let me know if we're on the right track here? Before I commit to developing for various url patterns I want to make sure we are going in the correct direction. I'll proably resume after the weekend as well, enjoy the 🎆 !!

**Comment by 23d1** [comment]
That is great! Sorry for the delay in commenting.
